alacritty/res/rect.f.glsl

```diff
@@ -9,46 +9,120 @@ flat in vec4 color;
 
 out vec4 FragColor;
 
-uniform int isUndercurl;
+uniform int rectKind;
 
 uniform float cellWidth;
 uniform float cellHeight;
 uniform float paddingY;
 uniform float paddingX;
 
-uniform float undercurlThickness;
+uniform float underlinePosition;
+uniform float underlineThickness;
+
 uniform float undercurlPosition;
 
+#define UNDERCURL 1
+#define DOTTED 2
+#define DASHED 3
+
 #define PI 3.1415926538
 
-void main()
-{
-    if (isUndercurl == 0) {
-        FragColor = color;
-        return;
-    }
+vec4 draw_undercurl(int x, int y) {
+  // We use `undercurlPosition` as an amplitude, since it's half of the descent
+  // value.
+  float undercurl =
+      -1. * undercurlPosition / 2. * cos(float(x) * 2 * PI / cellWidth) +
+      cellHeight - undercurlPosition;
 
-    int x = int(gl_FragCoord.x - paddingX) % int(cellWidth);
-    int y = int(gl_FragCoord.y - paddingY) % int(cellHeight);
+  float undercurlTop = undercurl + max((underlineThickness - 1), 0);
+  float undercurlBottom = undercurl - max((underlineThickness - 1), 0);
 
-    // We use `undercurlPosition` as amplitude, since it's half of the descent
-    // value.
-    float undercurl = -1. * undercurlPosition / 2.
-        * cos(float(x) * 2 * PI / float(cellWidth))
-        + cellHeight - undercurlPosition;
+  // Compute resulted alpha based on distance from `gl_FragCoord.y` to the
+  // cosine curve.
+  float alpha = 1.;
+  if (y > undercurlTop || y < undercurlBottom) {
+    alpha = 1. - min(abs(undercurlTop - y), abs(undercurlBottom - y));
+  }
 
-    // We subtract one, since curve is already 1px thick.
-    float undercurl_top = undercurl + max((undercurlThickness - 1), 0);
-    float undercurl_bottom = undercurl - max((undercurlThickness - 1), 0);
+  // The result is an alpha mask on a rect, which leaves only curve opaque.
+  return vec4(color.rgb, alpha);
+}
 
+// When the dot size increases we can use AA to make spacing look even and the
+// dots rounded when the size allows us.
```

Comment:
```suggestion
// When the dot size increases we can use AA to make spacing look even and the
// dots rounded.
```

Already mentioned in the beginning that we're talking about increased dot size, no need to say it twice.